### ISSUE TITLE
qemu-system-x86_64:/var/log/lxd/avm/qemu.conf:68: Could not open '/var/lib/lxd/virtual-machines/avm/qemu.nvram': No such file or directory

### ISSUE THREAD

**Issue by cvmiller** [lxc/lxd#8216]
# Required information

 * Distribution: Artix
 * Distribution version: Rolling release
 * The output of "lxc info" or if that fails:
   * Kernel version:
   * LXC version:
   * LXD version:
   * Storage backend in use:
cvmiller@koa:~$ lxc info
config:
  images.auto_update_interval: "0"
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
- resources_network_firmware
- backup_compression_algorithm
- ceph_data_pool_name
- container_syscall_intercept_mount
- compression_squashfs
- container_raw_mount
- container_nic_routed
- container_syscall_intercept_mount_fuse
- container_disk_ceph
- virtual-machines
- image_profiles
- clustering_architecture
- resources_disk_id
- storage_lvm_stripes
- vm_boot_priority
- unix_hotplug_devices
- api_filtering
- instance_nic_network
- clustering_sizing
- firewall_driver
- projects_limits
- container_syscall_intercept_hugetlbfs
- limits_hugepages
- container_nic_routed_gateway
- projects_restrictions
- custom_volume_snapshot_expiry
- volume_snapshot_scheduling
- trust_ca_certificates
- snapshot_disk_usage
- clustering_edit_roles
- container_nic_routed_host_address
- container_nic_ipvlan_gateway
- resources_usb_pci
- resources_cpu_threads_numa
- resources_cpu_core_die
- api_os
- container_nic_routed_host_table
- container_nic_ipvlan_host_table
- container_nic_ipvlan_mode
- resources_system
- images_push_relay
- network_dns_search
- container_nic_routed_limits
- instance_nic_bridged_vlan
- network_state_bond_bridge
- usedby_consistency
- custom_block_volumes
- clustering_failure_domains
- resources_gpu_mdev
- console_vga_type
- projects_limits_disk
- network_type_macvlan
- network_type_sriov
- container_syscall_intercept_bpf_devices
- network_type_ovn
- projects_networks
- projects_networks_restricted_uplinks
- custom_volume_backup
- backup_override_name
- storage_rsync_compression
- network_type_physical
- network_ovn_external_subnets
- network_ovn_nat
- network_ovn_external_routes_remove
- tpm_device_type
- storage_zfs_clone_copy_rebase
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIB+jCCAYGgAwIBAgIRAJCkgsvYLMXJkV2ARLL3KeswCgYIKoZIzj0EAwMwMTEc
    MBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzERMA8GA1UEAwwIcm9vdEBrb2Ew
    HhcNMjAxMTIwMDAyMTU1WhcNMzAxMTE4MDAyMTU1WjAxMRwwGgYDVQQKExNsaW51
    eGNvbnRhaW5lcnMub3JnMREwDwYDVQQDDAhyb290QGtvYTB2MBAGByqGSM49AgEG
    BSuBBAAiA2IABNcI9bb2clVd/j6c8vLzQ3NfzvVjPJKUrgJTbhw8zkJJx4MESiza
    cJm6J2FI/ySgQPujyOx2qhVI7diM4gqpxwebnvppp3tCHoo+PvcsLJ7bTqH1NKDs
    9HoXPK42SpBsRaNdMFswDgYDVR0PAQH/BAQDAgWgMBMGA1UdJQQMMAoGCCsGAQUF
    BwMBMAwGA1UdEwEB/wQCMAAwJgYDVR0RBB8wHYIDa29hhwR/AAABhxAAAAAAAAAA
    AAAAAAAAAAABMAoGCCqGSM49BAMDA2cAMGQCMH2TvwVotPxq6yoEs9KzUav6jW+W
    N5vmk+sd9Pj/OwIYibx5T7MmurGE8eSMk97s5QIwLny+JPhruEu9cFygXh1e2NF4
    G2BMnBkIwOC2Dywp1J7W08QMQ8Kx0RlTcVQO4mNb
    -----END CERTIFICATE-----
  certificate_fingerprint: 9f74d499873191ab76bf238dcb08c414fa32a58abc41f74f6351843117d55e68
  driver: lxc
  driver_version: 4.0.5
  firewall: xtables
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    seccomp_listener_continue: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.9.6-artix1-1
  lxc_features:
    cgroup2: "true"
    devpts_fd: "true"
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    network_veth_router: "true"
    pidfd: "true"
    seccomp_allow_deny_syntax: "true"
    seccomp_notify: "true"
    seccomp_proxy_send_notify_fd: "true"
  os_name: Artix Linux
  os_version: ""
  project: default
  server: lxd
  server_clustered: false
  server_name: koa
  server_pid: 7005
  server_version: "4.8"
  storage: btrfs
  storage_version: "5.9"

# Issue description

A brief description of the problem. Should include what you were
attempting to do, what you did, what happened and what you expected to
see happen.

When attempting to create a new Virtual Machine, the VM fails with the error message:
qemu-system-x86_64:/var/log/lxd/avm/qemu.conf:68: Could not open '/var/lib/lxd/virtual-machines/avm/qemu.nvram': No such file or directory

# Steps to reproduce

 1. Launch a vm `lxc launch -p extbridge images:alpine/edge avm --vm`
 2. Note that it fails
 3. examine log 
```
 lxc info --show-log local:avm
Name: avm
Location: none
Remote: unix://
Architecture: x86_64
Created: 2020/12/04 23:44 UTC
Status: Stopped
Type: virtual-machine
Profiles: extbridge

Log:

qemu-system-x86_64:/var/log/lxd/avm/qemu.conf:68: Could not open '/var/lib/lxd/virtual-machines/avm/qemu.nvram': No such file or directory

```

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

I have found that the file OVMF_VARS.fd is not being mapped correctly into the VM qemu directory `/var/lib/lxd/virtual-machines/avm`. Doing a `ls -l` shows that the symlink is incorrect:
```
koa:[root]:/var/lib/lxd/virtual-machines/avm# ls -l
total 4194352
-rw-r--r-- 1 root   root        692 Dec  4 15:44 agent-client.crt
-rw------- 1 root   root        288 Dec  4 15:44 agent-client.key
-rw-r--r-- 1 root   root        725 Dec  4 15:44 agent.crt
-rw------- 1 root   root        288 Dec  4 15:44 agent.key
-r-------- 1 root   root       1599 Dec  4 15:44 backup.yaml
dr-x------ 1 nobody root        180 Dec  4 15:44 config
-rw-r--r-- 1 root   root        660 Dec  4 05:02 metadata.yaml
lrwxrwxrwx 1 root   root         12 Dec  4 15:44 qemu.nvram -> OVMF_VARS.fd
-rw-r--r-- 1 root   root 9999998976 Dec  4 15:44 root.img
drwxr-xr-x 1 root   root         64 Dec  4 05:02 templates
```
A work-around, is to copy the OVMF_VARS.fd into this directory (/var/lib/lxd/virtual-machines/avm), and then restart the vm with the command `lxc start avm`

The OVMF_VARS.fd is located at: `/usr/share/OVMF/`

The VM will start normally after this.

**Comment by cvmiller** [comment]
Captured lxd debug log with command `sudo lxd --group lxd --debug` during the launch of the VM
[lxd.log](https://github.com/lxc/lxd/files/5646126/lxd.log)

**Comment by stgraber** [comment]
Ah, looks like we need to deref the source path of OVMF_VARS.fd at least to accommodate for that.

**Comment by monstermunchkin** [comment]
> The OVMF_VARS.fd is located at: /usr/share/OVMF/

@cvmiller Are you sure? The [ovmf package repo](https://gitea.artixlinux.org/artixlinux/packages/src/branch/master/ovmf/trunk/PKGBUILD) for Artix suggests that it's located at `/usr/share/ovmf/x64/OVMF_VARS.fd`. In that case, setting `LXD_OVMF_PATH` to `/usr/share/ovmf/x64` should do the trick.

**Comment by cvmiller** [comment]
It is true, Artix puts OVMF files in `/usr/share/ovmf/x64/OVMF_VARS.fd`, and I created a sym-link to that directory to make it appear to be in ` /usr/share/OVMF/`

I also tried to use ENV vars in the /etc/init.d/lxd script. But Artix is running OpenRC, and doesn't appear to pass ENV Vars down to children. Is there a way to put ENV vars in /etc/config.d/lxd ? I don't see this method in the LXD documentation.
